feat(sqllab): SQLEditor Extension

Antonio-RiveroMartnez opened the pull request:
### SUMMARY
In order to extend capabilities in our SQLEditor we need to provide an extension. This PR is adding such extension and adding tests for it.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

### TESTING INSTRUCTIONS
<!--- Required! What steps can be taken to manually verify the changes? -->

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

Antonio-RiveroMartnez committed:
SQLEditor extension:
    - Add a new extension so our SQLEditor can make use of it if provided extend SQLLab.
    - Add tests for the new extension

justinpark commented on superset-frontend/packages/superset-ui-core/src/ui-overrides/ExtensionsRegistry.ts:
```diff
@@ -109,6 +109,27 @@ export interface DatabaseConnectionExtension {
+  TQueryEditor = object,
+  TDatabase = any,
+> {
+  tables: TTables[];
+  queryEditor: TQueryEditor;
+  database: TDatabase;
```
@Antonio-RiveroMartnez 
Not all custom extension requires all these properties such like `tables` and `database`
Therefore, I think it's not a good idea passing these specified properties because you can access these values from redux. (you can also access the queryEditor object by useQueryEditor hook)
Only thing you need to pass is `queryEditorId` and then each custom extension component should use a hook to retrieve the properties that needed.

```suggestion
  queryEditorId: string;
```

Antonio-RiveroMartnez replied:
Hey @justinpark ! Good call, I will change this. I am working a bit more on this extension feature so I'll move this PR to be a DRAFT until I get everything in place and final props etc are figured out. ILYK. Thanks for the input as always 👏

michael-s-molina commented:
@Antonio-RiveroMartnez could you help me review https://github.com/apache/superset/pull/24214 and rebase your PR on top of that one?

Antonio-RiveroMartnez committed:
SQLEditor extension:

- Rebase using the refactored branch

Antonio-RiveroMartnez committed:
SQLEditor extension:

- Use queryEditor Id instead of passing tables and database and add two additional props needed for triggering queries

betodealmeida reviewed (approved):
Nice!

eschutho reviewed (commented):
LGTM!

justinpark commented on superset-frontend/src/SqlLab/components/SqlEditor/index.jsx:
```diff
@@ -673,6 +684,15 @@ const SqlEditor = ({
+          {SqlFormExtension && (
+            <SqlFormExtension
+              queryEditorId={queryEditor.id}
+              setQueryEditorAndSaveSqlWithDebounce={
```
@geido @Antonio-RiveroMartnez similar to [the previous comment](https://github.com/apache/superset/pull/24205#discussion_r1205807291), this action props doesn't make sense since your custom component only needs these actions. You can hook up these actions in your custom component by combining of existing action helpers

For example you can build startQuery in your component by combining of runQueryFromSqlEditor and setActiveSouthPaneTab
```
      dispatch(
        runQueryFromSqlEditor(
          database,
          queryEditor,
          defaultQueryLimit,
          ctasArg ? ctas : '',
          ctasArg,
          ctas_method,
        ),
      );
      dispatch(setActiveSouthPaneTab('Results'));
```

Antonio-RiveroMartnez replied:
Hey @justinpark sure, I can recreate all those with the helpers, but since we are not changing how they work, just invoking them, I treated them like other props our existing extensions are using. IMHO I see no major benefit by forcing a duplication of code in the extension. However, if you truly see this like an scenario we should force such duplication, I can work on it in a follow up PR. Thanks as always for the input.
